### ISSUE TITLE
MERGE and REPLACE problem/question in ConflictResolution  or bug with REPLACE (JAVA)

### ISSUE THREAD

**Issue by tihiydo** [anggrayudi/SimpleStorage#79]
Simple Storage 0.14.0

I have two folders

/movies/
/app/movies/

They have a lot of identical files, well, for example, there

/movies/filmAboutLove.mp4
/movies/horrorFilm.mp4

/app/movies/filmAboutLove.mp4
/app/movies/horrorFilm.mp4

I need update /app/movies/ folder and all files inside

That is, I need to replace, or add new files from the folder /movies/ to the folder /app/movies/

`
            

            DocumentFile f = DocumentFileCompat.fromFullPath(layoutContext, storage + "/movies");
            DocumentFile f2 = DocumentFileCompat.fromFullPath(layoutContext, storage  + "/app/movies");
            DocumentFileUtils.moveFolderTo(f, layoutContext, f2, true, null, new FolderCallback() {

            @Override
            public long onStart(DocumentFile param1DocumentFile, int param1Int, Thread workerThread)
            {
                return 1000L;
            }

            @Override
            public void onFailed(ErrorCode errorCode)
            {
            }

            @Override
            public void onParentConflict(DocumentFile param1DocumentFile, ParentFolderConflictAction action, boolean canMerge)
            {
                action.confirmResolution(ConflictResolution.MERGE);
            }

            @Override
            public void onContentConflict(DocumentFile destinationFolder, List<FileConflict> conflictedFiles, FolderContentConflictAction action)
            {
                handleFolderContentConflict(action, conflictedFiles);
            }

            public void handleFolderContentConflict(FolderContentConflictAction action, List<FileConflict> conflictedFiles)
            {
                askSolution(action, conflictedFiles, new ArrayList<FileConflict>(conflictedFiles.size()));
            }

            public void askSolution(FolderContentConflictAction action, List<FileConflict> conflictedFiles, List<FileConflict> newSolution)
            {
                FileConflict currentSolution = CollectionsKt.removeFirstOrNull(conflictedFiles);
                if (currentSolution == null)
                {
                    action.confirmResolution(newSolution);
                    return;
                }
                currentSolution.setSolution(FileCallback.ConflictResolution.REPLACE);
                conflictedFiles.forEach (it -> it.setSolution(currentSolution.getSolution()));
                newSolution.addAll(conflictedFiles);
                action.confirmResolution(newSolution);
            }

            @Override
            public void onReport(Report rep)
            {
                Log.d("DEBUG", String.format("Progress %f", rep.getProgress()));
            }

        });
    }
`
    
Valid if you use code above you will have in folder /app/movies/ this list of files

- filmAboutLove.mp4
- filmAboutLove (1).mp4
- horrorFilm.mp4
- horrorFilm (1).mp4

When it was necessary to simply replace existing files, or add new folder or files
But if we think about that example, then we should get

- filmAboutLove.mp4 (REPLACED)
- horrorFilm.mp4 (REPLACED)

In my code, a function with a replacement, namely **askSolution** not work.

I guess it's because of **ConflictResolution.MERGE**, but while it says in the description to this, we should have skipped duplicate files rather than doing **ConflictResolution.CREATE_NEW**.

But what worries me more is that I can't. Replace files in identical folders, while being able to create new files and folders.

**Comment by anggrayudi** [comment]
Will be fixed in `1.1.0`. For now, just use `1.1.0-SNAPSHOT`.

FYI, you are trying to replace all conflicted files without asking the user. Your `askSolution()` should be:
```java
public void askSolution(
        FolderCallback.FolderContentConflictAction action,
        List<FolderCallback.FileConflict> conflictedFiles,
        List<FolderCallback.FileConflict> newSolution
) {
    newSolution.addAll(conflictedFiles);
    newSolution.forEach(it -> it.setSolution(FileCallback.ConflictResolution.REPLACE));
    action.confirmResolution(newSolution);
}
```

**Comment by tihiydo** [comment]
Also, on version 0.14.0 of the library, I noticed a bug, if you use moveFolderTo and specify a custom new folder name, then only files without directories will be transferred. Tried to contact you and inform through the social. networks. about it but couldn't.

Perhaps this has already been fixed in the new versions, but I decided to inform you all of a sudden that

`
BUG
DocumentFileUtils.moveFolderTo(f, layoutContext, f2, true, {CUSTOM_NAME}, new FolderCallback() { }`

Now I will try updates with unpacking archives

**Comment by tihiydo** [comment]
Hmm, i can't use snapshot, repo it's not work

**Comment by anggrayudi** [comment]
Wrong implementation of the Maven snapshot URL. Please read this: https://github.com/anggrayudi/SimpleStorage#overview

**Comment by anggrayudi** [comment]
> if you use moveFolderTo and specify a custom new folder name, then only files without directories will be transferred.

Also fixed in `1.1.0-SNAPSHOT`. Please re-download/refresh your dependencies.

**Comment by anggrayudi** [comment]
[1.1.0](https://github.com/anggrayudi/SimpleStorage/releases/tag/1.1.0) has been released.

**Comment by tihiydo** [comment]
> [1.1.0](https://github.com/anggrayudi/SimpleStorage/releases/tag/1.1.0) has been released.

I described u maybe new bug in tg, with copyFile

**Comment by anggrayudi** [comment]
The bug in `copyFile()` also fixed in 1.0.0

**Comment by tihiydo** [comment]
> The bug in `copyFile()` also fixed in 1.0.0

1.1.0-SNAPSHOT don't work copyFile (without errors)
I will send you code. I try to backup file, if backup file it's not exist

```
            public void handleFolderContentConflict(FolderContentConflictAction action, List<FileConflict> conflictedFiles) {
                askSolution(action, conflictedFiles, new ArrayList<FileConflict>(conflictedFiles.size()));
            }

            private void askSolution(FolderContentConflictAction action, List<FileConflict> conflictedFiles, List<FileConflict> newSolution) {
                FileConflict currentSolution = CollectionsKt.removeFirstOrNull(conflictedFiles);
                if (currentSolution == null) {
                    action.confirmResolution(newSolution);
                    return;
                }

                DocumentFile BackUpFileFolder = DocumentFileUtils.findParent(currentSolution.getTarget(), layoutContext);
                if ((DocumentFileUtils.findFileLiterally(BackUpFileFolder, currentSolution.getTarget().getName() + ".bm")) == null)
                {
                    Log.d("zxc", "This works");
                    DocumentFileUtils.copyFileTo(currentSolution.getTarget(), layoutContext, BackUpFileFolder, new FileDescription(currentSolution.getTarget().getName() + ".bm"), new FileCallback() { });
                    Log.d("zxc", "This not work, onConflict empty");
                }
                currentSolution.setSolution(FileCallback.ConflictResolution.REPLACE);
                newSolution.add(currentSolution);

                askSolution(action, conflictedFiles, newSolution);
            }
        });
    }
```

nothing happens on the copyFile To function and the next log stops
as if something does not let him through, waiting for Promise
I tried to display onConflict, onFailed but it is not even called
maybe a new asynchronous task is needed not to write everything in one

**Comment by anggrayudi** [comment]
I have tested `copyFileTo()` and it works. Your implementation is wrong. Putting `DocumentFileUtils.copyFileTo()` in the `DocumentFileUtils.moveFolderTo()`'s `onContentConflict()` may suspend the thread forever.
